File: hashing/contiguous_array.java
public int findMaxLengthDS1(int[] nums)
{
    int max_len = 0;

    for (int start = 0; start < nums.length; start++)
    {
        int zeros = 0;
        int ones = 0;

        for (int end = start; end < nums.length; end++)
        {
            if (nums[end] == 0)
            {
                zeros++;
            }
            else
            {
                ones++;
            }

            if (zeros == ones)
            {
                max_len = Math.max(max_len, end - start + 1);
            }
        }
    }

    return max_len;
}

public int findMaxLengthDS2(int[] nums)
{
    int nums_len = nums.length;

    int[] arr_of_counts = new int[2 * nums_len + 1];
    Arrays.fill(arr_of_counts, -2);
    arr_of_counts[nums_len] = -1;

    int max_len = 0;
    int count = 0;

    for (int idx = 0; idx < nums_len; idx++)
    {
        count = count + (nums[idx] == 0 ? -1 : 1);

        int idx_in_2n_plus_1 = count + nums_len;
        if (arr_of_counts[idx_in_2n_plus_1] >= -1)
        {
            max_len = Math.max(max_len, idx - arr_of_counts[idx_in_2n_plus_1]);
        }
        else
        {
            arr_of_counts[idx_in_2n_plus_1] = idx;
        }
    }

    return max_len;
}

public int findMaxLengthDS3(int[] nums)
{
    Map<Integer, Integer> count_idx = new HashMap<>();
    count_idx.put(0, -1);

    int max_len = 0;
    int count = 0;

    for (int idx = 0; idx < nums.length; idx++)
    {
        count = count + (nums[idx] == 0 ? -1 : 1);

        if (count_idx.containsKey(count))
        {
            max_len = Math.max(max_len, idx - count_idx.get(count));
        }
        else
        {
            count_idx.put(count, idx);
        }
    }

    return max_len;
}

File: hashing/contiguous_array.py
def findMaxLengthDS1(nums: list[int]) -> int:
    max_len = 0
    nums_len = len(nums)

    for start in range(nums_len):
        zeros = 0
        ones = 0

        for end in range(start, nums_len):
            if nums[end] == 0:
                zeros += 1
            else:
                ones += 1

            if zeros == ones:
                max_len = max(max_len, end - start + 1)

    return max_len


def findMaxLengthDS2(nums: list[int]) -> int:
    nums_len = len(nums)
    arr_of_counts = [-2] * (2 * nums_len + 1)
    arr_of_counts[nums_len] = -1

    max_len = 0
    count = 0

    for idx in range(nums_len):
        count += -1 if nums[idx] == 0 else 1

        idx_in_2n_plus_1 = count + nums_len
        if arr_of_counts[idx_in_2n_plus_1] >= -1:
            max_len = max(max_len, idx - arr_of_counts[idx_in_2n_plus_1])
        else:
            arr_of_counts[idx_in_2n_plus_1] = idx

    return max_len


def findMaxLengthDS3(nums: list[int]) -> int:
    count_idx = {0: -1}

    max_len = 0
    count = 0

    for idx in range(len(nums)):
        count += (-1 if nums[idx] == 0 else 1)

        if count in count_idx:
            max_len = max(max_len, idx - count_idx[count])
        else:
            count_idx[count] = idx

    return max_len
